core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala

```diff
@@ -694,7 +694,7 @@ private[storage] class PartiallyUnrolledIterator[T](
   }
 
   override def next(): T = {
-    if (unrolled == null) {
+    if (unrolled == null || !unrolled.hasNext) {
```

Comment:
I agree that next() without hasNext is a valid flow. However, the caller which behaves like that should also aware of the possibility of no element exception. 

`TorrentBroadcast` is problematic because it doesn't call hasNext and doesn't handle this possibility.

I'd prefer to revert the added `hasNext` call in `next`. But not strong option.